cmd/kubeapps-apis/core/packages/v1alpha1/packages_fan_in.go

```diff
@@ -0,0 +1,217 @@
+// that the caller can generate a next page token that is able to encode the
+// offsets for each plugin. The next request the begins each plugin where it
+// left off for the last.
+//
+// Plugins generally do not use snapshots of the actual data, so, similar to the
+// pagination of individual plugins, it will be possible that this returns
+// duplicates or missing data if data is added or removed between paginated
+// requests.
+func fanInAvailablePackageSummaries(ctx context.Context, pkgPlugins []pkgPluginWithServer, request *packages.GetAvailablePackageSummariesRequest) (<-chan summaryWithOffsets, error) {
+	summariesCh := make(chan summaryWithOffsets)
+
+	corePageSize := int(request.GetPaginationOptions().GetPageSize())
+	// We'll request a bit more than pageSize / n from each plugin. If the page
+	// size is 10 and we have 3 plugins, request 5 items from each to start.
+	pluginPageSize := corePageSize
+	if len(pkgPlugins) > 1 {
+		pluginPageSize = pluginPageSize / (len(pkgPlugins) - 1)
+	}
+
+	pluginPageOffsets := map[string]int{}
+	if request.GetPaginationOptions().GetPageToken() != "" {
+		err := json.Unmarshal([]byte(request.GetPaginationOptions().GetPageToken()), &pluginPageOffsets)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal %q: %w", request.GetPaginationOptions().GetPageToken(), err)
+		}
+	}
+
+	fanInput := []<-chan *summaryWithOffset{}
+	for _, pluginWithSrv := range pkgPlugins {
+		// Importantly, each plugin needs its own request, with its on pagination
+		// options.
+		r := &packages.GetAvailablePackageSummariesRequest{
+			Context:       request.Context,
+			FilterOptions: request.FilterOptions,
+			PaginationOptions: &packages.PaginationOptions{
+				PageSize:  int32(pluginPageSize),
+				PageToken: fmt.Sprintf("%d", pluginPageOffsets[pluginWithSrv.plugin.Name]),
+			},
+		}
+
+		ch, err := sendAvailablePackageSummariesForPlugin(ctx, pluginWithSrv, r)
+		if err != nil {
+			return nil, err
+		}
+		fanInput = append(fanInput, ch)
+	}
+
+	// We now have a slice of channels for the fan-in and want a go routine that
+	// will ensure it sends the next (ordered) item from all channels down the
+	// channel.
+	go func() {
+		numSent := 0
+		nextItems := make([]*summaryWithOffset, len(fanInput))
+		for {
+			// Populate the empty next items from each channel.
+			for i, ch := range fanInput {
+				if nextItems[i] == nil {
+					// If the channel is closed, the value will remain nil.
+					ok := true
+					nextItems[i], ok = <-ch
+					if !ok {
+						// If the channel was closed, we reached the last item for that
+						// plugin. We need to recognise when all plugins have exhausted
+						// itemsoffsets
+						pluginName := pkgPlugins[i].plugin.Name
+						pluginPageOffsets[pluginName] = CompleteToken
+					}
+
+					if nextItems[i] != nil && nextItems[i].err != nil {
+						summariesCh <- summaryWithOffsets{
+							err: nextItems[i].err,
+						}
+						close(summariesCh)
+						return
+					}
+				}
```

Comment:
So this part ensures that `nextItems` is populated for each plugin (that hasn't yet been exhausted)...